node/src/components/consensus/protocols/highway/round_success_meter.rs

```diff
@@ -0,0 +1,146 @@
+use std::{cmp::max, collections::VecDeque, mem};
+
+use datasize::DataSize;
+
+use crate::{
+    components::consensus::{
+        highway_core::{finality_detector::FinalityDetector, State, Weight},
+        traits::Context,
+    },
+    types::Timestamp,
+};
+
+const NUM_ROUNDS_TO_CONSIDER: usize = 40;
+const MAX_FAILED_ROUNDS: usize = 10;
+const ACCELERATION_PARAMETER: u64 = 1000;
+const THRESHOLD: u64 = 1;
+
+#[derive(DataSize, Debug)]
+pub(crate) struct RoundSuccessMeter<C>
+where
+    C: Context,
+{
+    // store whether a particular round was successful
+    // index 0 is the last handled round, 1 is the second-to-last etc.
+    rounds: VecDeque<bool>,
+    current_round_id: u64,
+    proposals: Vec<C::Hash>,
+    min_round_exp: u8,
+    current_round_exp: u8,
+    current_round_length: u64,
+}
+
+impl<C: Context> RoundSuccessMeter<C> {
+    pub fn new(round_exp: u8, min_round_exp: u8, timestamp: Timestamp) -> Self {
+        let current_round_length = 1u64 << round_exp;
+        let current_round_index = timestamp.millis() / current_round_length;
+        let current_round_id = current_round_index * current_round_length;
+        Self {
+            rounds: VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER),
+            current_round_id,
+            proposals: Vec::new(),
+            min_round_exp,
+            current_round_exp: round_exp,
+            current_round_length,
+        }
+    }
+
+    fn change_exponent(&mut self, new_exp: u8, timestamp: Timestamp) {
+        self.rounds = VecDeque::with_capacity(NUM_ROUNDS_TO_CONSIDER);
+        self.current_round_exp = new_exp;
+        self.current_round_length = 1u64 << new_exp;
+        let round_index = timestamp.millis() / self.current_round_length;
+        self.current_round_id = round_index * self.current_round_length;
+        self.proposals = Vec::new();
+    }
+
+    fn check_proposals_success(&self, state: &State<C>, proposal_h: &C::Hash) -> bool {
+        let total_w = state.total_weight();
+
+        let finality_detector =
+            FinalityDetector::<C>::new(max(total_w / 100 * THRESHOLD, Weight(1)));
```

Comment:
Shouldn't we first multiply `total_w * THRESHOLD` and only then divide? Otherwise, if `total_w < 100` we will end up with `0`.

Reply:
See Andreas' comment above. I also thought that at first, but I guess this is the question of whether we worry more about overflows or underflows. With `THRESHOLD` set to 1 this doesn't even make a difference, though.